pkg/schedule/affinity/labeler.go

```diff
@@ -0,0 +1,709 @@
+			newRange := keyRange{
+				StartKey: addOp.StartKey,
+				EndKey:   addOp.EndKey,
+				GroupID:  groupID,
+			}
+			currentRanges = append(currentRanges, newRange)
+			allNewRanges = append(allNewRanges, newRange)
+		}
+
+		updatedRanges[groupID] = currentRanges
+	}
+
+	// Validate no overlaps with newly added ranges
+	if len(allNewRanges) > 0 {
+		if err := m.validateNoKeyRangeOverlap(allNewRanges); err != nil {
+			return err
+		}
+	}
+
+	// Update label rules and cache for all affected groups
+	for groupID, ranges := range updatedRanges {
+		if err := m.updateGroupRanges(groupID, ranges); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getCurrentRanges retrieves the current key ranges for a group.
+func (m *Manager) getCurrentRanges(groupID string) ([]keyRange, error) {
+	// Try cache first
+	if ranges := m.keyRanges[groupID]; ranges != nil {
+		return append([]keyRange(nil), ranges...), nil
+	}
+
+	// Parse from label rule
+	labelRule := m.regionLabeler.GetLabelRule(GetLabelRuleID(groupID))
+	if labelRule == nil {
+		return nil, errors.Errorf("label rule not found for group %s", groupID)
+	}
+
+	dataSlice, ok := labelRule.Data.([]*labeler.KeyRangeRule)
+	if !ok {
+		return nil, errors.Errorf("invalid label rule data type for group %s, got type %T", groupID, labelRule.Data)
+	}
+
+	return parseKeyRangesFromData(dataSlice, groupID)
+}
+
+// applyRemoveOps filters out ranges that match remove operations.
+// Optimized with a map for O(n+m) complexity instead of O(n*m).
+func applyRemoveOps(currentRanges []keyRange, removes []GroupRangeModification) []keyRange {
+	if len(removes) == 0 {
+		return currentRanges
+	}
+
+	// Build a set of ranges to remove for O(1) lookup
+	// Use hex encoding to avoid key collisions
+	removeSet := make(map[string]struct{}, len(removes))
+	for _, r := range removes {
+		key := hex.EncodeToString(r.StartKey) + "|" + hex.EncodeToString(r.EndKey)
+		removeSet[key] = struct{}{}
+	}
+
+	var filtered []keyRange
+	for _, current := range currentRanges {
+		key := hex.EncodeToString(current.StartKey) + "|" + hex.EncodeToString(current.EndKey)
+		if _, found := removeSet[key]; !found {
+			filtered = append(filtered, current)
+		}
+	}
+	return filtered
+}
+
+// updateGroupRanges updates the label rule and cache for a group's key ranges.
+func (m *Manager) updateGroupRanges(groupID string, ranges []keyRange) error {
+	labelRule := m.regionLabeler.GetLabelRule(GetLabelRuleID(groupID))
+	if labelRule == nil {
+		return errors.Errorf("label rule not found for group %s", groupID)
+	}
+
+	// Allow clearing ranges by deleting the label rule and cache.
+	if len(ranges) == 0 {
+		if err := m.regionLabeler.DeleteLabelRule(labelRule.ID); err != nil {
+			return err
+		}
+		delete(m.keyRanges, groupID)
+		if groupInfo, ok := m.groups[groupID]; ok {
+			groupInfo.LabelRule = nil
+			groupInfo.RangeCount = 0
+			groupInfo.AffinityRegionCount = 0
+			m.affinityRegionCount -= groupInfo.AffinityRegionCount
```

Comment:
```suggestion
			m.affinityRegionCount -= groupInfo.AffinityRegionCount
			groupInfo.AffinityRegionCount = 0
```
Otherwise, m.affinityRegionCount will not always be changed.